File: pipeline/compare_methods.py
"""Compare area-weighted vs parcel-weighted income per civic association."""
from __future__ import annotations

import geopandas as gpd
import pandas as pd

from pipeline import config


def compare_income(area: pd.DataFrame, parcel: pd.DataFrame) -> pd.DataFrame:
    """Join the two methods' mean income; add diff (parcel - area) and pct_diff."""
    a = area[["geoid", "mean_income"]].rename(columns={"mean_income": "mean_income_area"})
    p = parcel[["geoid", "mean_income"]].rename(columns={"mean_income": "mean_income_parcel"})
    out = a.merge(p, on="geoid", how="outer")
    out["diff"] = out["mean_income_parcel"] - out["mean_income_area"]
    out["pct_diff"] = 100 * out["diff"] / out["mean_income_area"].where(
        out["mean_income_area"] > 0
    )
    return out


def main() -> None:
    civ = gpd.read_file(config.CIVIC_ASSOC)[["geoid", "region_name", "geometry"]]
    area = pd.read_csv(config.CIVIC_INCOME, dtype={"geoid": str})
    parcel = pd.read_csv(config.CIVIC_INCOME_PARCELS, dtype={"geoid": str})
    cmp = compare_income(area, parcel)
    gdf = civ.merge(cmp, on="geoid", how="left")
    gdf = gpd.GeoDataFrame(gdf, geometry="geometry", crs=civ.crs)

    # Honest accounting: report parcel-method household total vs ACS source.
    acs = gpd.read_file(config.ACS_COUNTS)
    src_hh = float(acs["households"].sum())
    parcel_hh = float(pd.read_csv(config.CIVIC_INCOME_PARCELS)["households"].sum())
    pct = 100 * (parcel_hh - src_hh) / src_hh if src_hh else 0.0
    print(f"Household totals, ACS source: {src_hh:,.0f}; parcel method: {parcel_hh:,.0f} ({pct:+.1f}%)")

    gdf.to_file(config.CIVIC_INCOME_COMPARISON, driver="GeoJSON")
    print(f"Wrote {len(gdf)} comparison rows → {config.CIVIC_INCOME_COMPARISON}")


if __name__ == "__main__":
    main()

File: pipeline-r/compare_methods.R
# Compare area-weighted vs parcel-weighted income per civic association.
# Parallels pipeline/compare_methods.py.
source("pipeline-r/config.R")
suppressPackageStartupMessages(library(sf))

compare_income <- function(area, parcel) {
  a <- data.frame(geoid = as.character(area$geoid),
                  mean_income_area = area$mean_income, stringsAsFactors = FALSE)
  p <- data.frame(geoid = as.character(parcel$geoid),
                  mean_income_parcel = parcel$mean_income, stringsAsFactors = FALSE)
  out <- merge(a, p, by = "geoid", all = TRUE)
  out$diff <- out$mean_income_parcel - out$mean_income_area
  out$pct_diff <- ifelse(out$mean_income_area > 0,
                         100 * out$diff / out$mean_income_area, NA_real_)
  out
}

run_compare <- function() {
  civ    <- sf::st_read(CIVIC_ASSOC, quiet = TRUE)[, c("geoid", "region_name")]
  civ$geoid <- as.character(civ$geoid)
  area   <- read.csv(CIVIC_INCOME_R,         colClasses = c(geoid = "character"))
  parcel <- read.csv(CIVIC_INCOME_PARCELS_R, colClasses = c(geoid = "character"))
  cmp <- compare_income(area, parcel)
  out <- merge(civ, cmp, by = "geoid", all.x = TRUE)
  sf::st_write(out, CIVIC_INCOME_COMPARISON_R, delete_dsn = TRUE, quiet = TRUE)
  cat(sprintf("Wrote %d comparison rows -> %s\n", nrow(out), CIVIC_INCOME_COMPARISON_R))
}

if (sys.nframe() == 0) run_compare()
